File: cli/cmd/history_graph_test.go
package cmd

import (
	"reflect"
	"testing"

	"github.com/kurrent-io/gaffer/cli/internal/remote"
)

// graphOf lays out a timeline from a newest-first list of content hashes. An empty
// string is a state change (no content identity), matching a row whose hash column
// is blank.
func graphOf(hashes ...string) historyGraph {
	vs := make([]historyVersion, len(hashes))
	for i, h := range hashes {
		vs[i].ContentHash = h
		vs[i].Hash = h
	}
	return computeHistoryGraph(vs)
}

func TestHistoryGraphNoReverts(t *testing.T) {
	g := graphOf("aaa", "bbb", "ccc")
	if len(g.spans) != 0 {
		t.Fatalf("spans = %+v, want none", g.spans)
	}
	if want := []int{0, 0, 0}; !reflect.DeepEqual(g.nodeLane, want) {
		t.Errorf("nodeLane = %v, want %v", g.nodeLane, want)
	}
}

func TestHistoryGraphSimpleRevert(t *testing.T) {
	// aaa reappears at row 0, matching row 3; rows 1-2 are the detour.
	g := graphOf("aaa", "bbb", "", "aaa")
	if len(g.spans) != 1 {
		t.Fatalf("spans = %+v, want one", g.spans)
	}
	if s := g.spans[0]; s.top != 0 || s.bottom != 3 || s.lane != 1 {
		t.Errorf("span = %+v, want {0 3 1}", s)
	}
	if want := []int{0, 1, 1, 0}; !reflect.DeepEqual(g.nodeLane, want) {
		t.Errorf("nodeLane = %v, want %v (endpoints on main, detour in lane 1)", g.nodeLane, want)
	}
}

func TestHistoryGraphAdjacentRewriteIsNotABranch(t *testing.T) {
	// Two identical writes with nothing between: a rewrite, not a revert.
	g := graphOf("aaa", "aaa", "bbb")
	if len(g.spans) != 0 {
		t.Fatalf("spans = %+v, want none (adjacent identical is a rewrite)", g.spans)
	}
}

func TestHistoryGraphSequentialRevertsShareALane(t *testing.T) {
	// aaa at rows 0, 2, 5: two back-to-back reverts sharing the middle endpoint.
	// Both are top-level, so both draw in lane 1.
	g := graphOf("aaa", "x", "aaa", "y", "z", "aaa")
	if len(g.spans) != 2 {
		t.Fatalf("spans = %+v, want two sequential", g.spans)
	}
	for _, s := range g.spans {
		if s.lane != 1 {
			t.Errorf("span %+v lane = %d, want 1 (sequential, not nested)", s, s.lane)
		}
	}
	if want := []int{0, 1, 0, 1, 1, 0}; !reflect.DeepEqual(g.nodeLane, want) {
		t.Errorf("nodeLane = %v, want %v", g.nodeLane, want)
	}
}

func TestHistoryGraphNestedRevert(t *testing.T) {
	// Outer aaa (0..6) with an inner bbb revert (2..4) inside its detour.
	g := graphOf("aaa", "d", "bbb", "e", "bbb", "f", "aaa")
	outer, inner := false, false
	for _, s := range g.spans {
		switch {
		case s.top == 0 && s.bottom == 6:
			outer = true
			if s.lane != 1 {
				t.Errorf("outer lane = %d, want 1", s.lane)
			}
		case s.top == 2 && s.bottom == 4:
			inner = true
			if s.lane != 2 {
				t.Errorf("inner lane = %d, want 2", s.lane)
			}
		}
	}
	if !outer || !inner {
		t.Fatalf("spans = %+v, want nested outer{0 6} + inner{2 4}", g.spans)
	}
	// Row 3 sits in the inner detour (lane 2); the inner endpoints (2, 4) sit in the
	// outer detour (lane 1); the outer endpoints (0, 6) on the main line.
	if want := []int{0, 1, 1, 2, 1, 1, 0}; !reflect.DeepEqual(g.nodeLane, want) {
		t.Errorf("nodeLane = %v, want %v", g.nodeLane, want)
	}
}

func TestHistoryGraphNestingCapDropsThirdLevel(t *testing.T) {
	// Three levels deep: outer (0..8), middle (2..6), inner (4..... wait
	// aaa 0/8, bbb 2/6, ccc 4/... needs a matching pair one level deeper.
	g := graphOf("aaa", "p", "bbb", "q", "ccc", "r", "ccc", "s", "bbb", "t", "aaa")
	//            0     1    2      3    4      5    6      7    8      9    10
	// aaa 0/10 (lane1), bbb 2/8 (lane2), ccc 4/6 would be lane3 -> dropped (flat).
	for _, s := range g.spans {
		if s.top == 4 && s.bottom == 6 {
			t.Fatalf("innermost span %+v was accepted, want dropped at the lane cap", s)
		}
		if s.lane >= historyMaxLanes {
			t.Errorf("span %+v lane %d exceeds cap %d", s, s.lane, historyMaxLanes)
		}
	}
	// Row 5 (the would-be inner detour) falls back to its enclosing lane (2), not 3.
	if g.nodeLane[5] >= historyMaxLanes {
		t.Errorf("nodeLane[5] = %d, want < %d (flattened)", g.nodeLane[5], historyMaxLanes)
	}
}

func TestHistoryGraphInterleaveNewestWins(t *testing.T) {
	// A at rows 0,2; B at rows 1,3 - the spans cross. The newer-top span (A, top 0)
	// keeps its bracket; the older (B, top 1) is dropped to flat.
	g := graphOf("A", "B", "A", "B")
	if len(g.spans) != 1 {
		t.Fatalf("spans = %+v, want one (the newer span wins the crossing)", g.spans)
	}
	if s := g.spans[0]; s.top != 0 || s.bottom != 2 {
		t.Errorf("kept span = %+v, want the newer {0 2}", s)
	}
	// A's detour (row 1) draws in lane 1; B's rows (1 already placed, 3) stay off any
	// bracket of their own - row 3 is on the main line.
	if want := []int{0, 1, 0, 0}; !reflect.DeepEqual(g.nodeLane, want) {
		t.Errorf("nodeLane = %v, want %v", g.nodeLane, want)
	}
}

func TestHistoryGraphEmpty(t *testing.T) {
	if g := graphOf(); len(g.spans) != 0 || len(g.nodeLane) != 0 {
		t.Errorf("empty history graph = %+v, want zero", g)
	}
}

func TestHistoryGraphIgnoresStateChangeEndpoints(t *testing.T) {
	// deploy b, disable (still b), deploy b: the content never diverged, so a bracket
	// here would be spurious - a state change can't anchor a revert.
	g := computeHistoryGraph(classifyHistory([]remote.Version{
		ver(2, "b", true, gafferLedger(remote.OpDeploy)),
		ver(1, "b", false, nil), // disabled - same content
		ver(0, "b", true, gafferLedger(remote.OpDeploy)),
	}))
	if len(g.spans) != 0 {
		t.Fatalf("spans = %+v, want none (a lifecycle toggle between identical deploys isn't a revert)", g.spans)
	}

	// deploy X, deploy Y, disable X, deploy X: a genuine revert. The endpoints are the
	// X deploys (rows 0 and 3), not the disabled row.
	g = computeHistoryGraph(classifyHistory([]remote.Version{
		ver(3, "X", true, gafferLedger(remote.OpDeploy)),
		ver(2, "Y", true, gafferLedger(remote.OpDeploy)),
		ver(1, "X", false, nil), // disabled X
		ver(0, "X", true, gafferLedger(remote.OpDeploy)),
	}))
	if len(g.spans) != 1 || g.spans[0].top != 0 || g.spans[0].bottom != 3 {
		t.Fatalf("spans = %+v, want one {0 3} anchored on the X deploys", g.spans)
	}
}

File: editors/vscode/src/panels/history-graph.test.ts
import { describe, expect, it } from "vitest";
import type { HistoryEntry } from "../commands/history-schema.js";
import {
	collapseHistory,
	computeHistoryGraph,
	HISTORY_MAX_LANES,
	type GraphNode,
} from "./history-graph.js";

// graphOf lays out a timeline from a newest-first list of content hashes. An
// empty string is a state change (no content identity) - matching cli/cmd's
// graphOf test helper.
function graphOf(...hashes: string[]) {
	return computeHistoryGraph(hashes.map((h) => ({ contentHash: h })));
}

describe("computeHistoryGraph", () => {
	it("no reverts: every node on the main line", () => {
		const g = graphOf("aaa", "bbb", "ccc");
		expect(g.spans).toHaveLength(0);
		expect(g.nodeLane).toEqual([0, 0, 0]);
	});

	it("simple revert: endpoints on main, detour in lane 1", () => {
		// aaa reappears at row 0, matching row 3; rows 1-2 are the detour.
		const g = graphOf("aaa", "bbb", "", "aaa");
		expect(g.spans).toEqual([{ top: 0, bottom: 3, lane: 1 }]);
		expect(g.nodeLane).toEqual([0, 1, 1, 0]);
	});

	it("adjacent identical write is a rewrite, not a branch", () => {
		const g = graphOf("aaa", "aaa", "bbb");
		expect(g.spans).toHaveLength(0);
	});

	it("sequential reverts sharing an endpoint share a lane", () => {
		// aaa at rows 0, 2, 5: two back-to-back reverts, both top-level in lane 1.
		const g = graphOf("aaa", "x", "aaa", "y", "z", "aaa");
		expect(g.spans).toHaveLength(2);
		for (const s of g.spans) expect(s.lane).toBe(1);
		expect(g.nodeLane).toEqual([0, 1, 0, 1, 1, 0]);
	});

	it("nested revert: inner detour one lane deeper", () => {
		// Outer aaa (0..6) with an inner bbb revert (2..4) inside its detour.
		const g = graphOf("aaa", "d", "bbb", "e", "bbb", "f", "aaa");
		expect(g.spans).toContainEqual({ top: 0, bottom: 6, lane: 1 });
		expect(g.spans).toContainEqual({ top: 2, bottom: 4, lane: 2 });
		expect(g.nodeLane).toEqual([0, 1, 1, 2, 1, 1, 0]);
	});

	it("nesting cap drops the third level to flat", () => {
		// aaa 0/10 (lane1), bbb 2/8 (lane2), ccc 4/6 would be lane3 -> dropped.
		const g = graphOf(
			"aaa",
			"p",
			"bbb",
			"q",
			"ccc",
			"r",
			"ccc",
			"s",
			"bbb",
			"t",
			"aaa",
		);
		expect(g.spans.some((s) => s.top === 4 && s.bottom === 6)).toBe(false);
		for (const s of g.spans) expect(s.lane).toBeLessThan(HISTORY_MAX_LANES);
		expect(g.nodeLane[5]).toBeLessThan(HISTORY_MAX_LANES);
	});

	it("interleaving spans: the newer-top span wins the crossing", () => {
		// A at rows 0,2; B at rows 1,3 - the spans cross. A (top 0) keeps its bracket.
		const g = graphOf("A", "B", "A", "B");
		expect(g.spans).toEqual([{ top: 0, bottom: 2, lane: 1 }]);
		expect(g.nodeLane).toEqual([0, 1, 0, 0]);
	});

	it("empty history is an empty graph", () => {
		const g = graphOf();
		expect(g.spans).toHaveLength(0);
		expect(g.nodeLane).toHaveLength(0);
	});

	it("a state change can't anchor a revert", () => {
		// deploy b, disable (still b), deploy b: the content never diverged.
		const noDetour: GraphNode[] = [
			{ contentHash: "b" },
			{ contentHash: "b", stateChange: true },
			{ contentHash: "b" },
		];
		expect(computeHistoryGraph(noDetour).spans).toHaveLength(0);

		// deploy X, deploy Y, disable X, deploy X: a genuine revert on the X deploys.
		const realDetour: GraphNode[] = [
			{ contentHash: "X" },
			{ contentHash: "Y" },
			{ contentHash: "X", stateChange: true },
			{ contentHash: "X" },
		];
		expect(computeHistoryGraph(realDetour).spans).toEqual([
			{ top: 0, bottom: 3, lane: 1 },
		]);
	});
});

function entry(version: number, kind: string): HistoryEntry {
	return {
		version,
		time: "",
		contentHash: "",
		kind,
		enabled: false,
		external: false,
		stateChange: false,
		deleted: false,
	};
}

describe("collapseHistory", () => {
	it("folds a recreate's delete + disable bookends into the recreate row", () => {
		const rows = collapseHistory([
			entry(5, "recreate"),
			entry(4, "deleted"),
			entry(3, "disabled"),
			entry(2, "deploy"),
		]);
		expect(rows.map((r) => r.version)).toEqual([5, 2]);
		expect(rows[0]?.absorbed.map((a) => a.version)).toEqual([4, 3]);
		expect(rows[1]?.absorbed).toHaveLength(0);
	});

	it("folds delete + rewritten (already-disabled recreate)", () => {
		const rows = collapseHistory([
			entry(3, "recreate"),
			entry(2, "deleted"),
			entry(1, "rewritten"),
		]);
		expect(rows).toHaveLength(1);
		expect(rows[0]?.absorbed.map((a) => a.version)).toEqual([2, 1]);
	});

	it("folds only the delete when the next row isn't a disable/rewrite", () => {
		const rows = collapseHistory([
			entry(3, "recreate"),
			entry(2, "deleted"),
			entry(1, "deploy"),
		]);
		expect(rows.map((r) => r.version)).toEqual([3, 1]);
		expect(rows[0]?.absorbed.map((a) => a.version)).toEqual([2]);
	});

	it("does not fold a recreate without an adjacent delete", () => {
		const rows = collapseHistory([entry(2, "recreate"), entry(1, "disabled")]);
		expect(rows.map((r) => r.version)).toEqual([2, 1]);
		expect(rows[0]?.absorbed).toHaveLength(0);
	});
});
